### PR TITLE
fix: Allow embedded manifests in all commands 

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This is a part of #12207.

One of the goals is for embedded manifests to be a first class citizen.  If you have a script, you should be able to run tests on it, for example.

This expands the error check from just `Cargo.toml` to also single-file packages so you can use it in `--manifest-path`.

This, however, does mean that these *can* be used in places that likely won't work yet, like `cargo publish`.

### How should we test and review this PR?

By commit.  We introduce tests for basic commands and then implement and refine the support for this.

### Additional information

Other information you want to mention in this PR, such as prior arts,
future extensions, an unresolved problem, or a TODO list.

### CODE REVIEW

**Review by weihanglo — COMMENTED** [review]
Looks good. Just wanting more tests to cover these error paths.

File: src/cargo/util/toml/mod.rs [hunk 1]
```diff
@@ -75,9 +75,11 @@ pub fn read_manifest(
 }
 
 /// See also `bin/cargo/commands/run.rs`s `is_manifest_command`
-fn is_embedded(path: &Path) -> bool {
+pub fn is_embedded(path: &Path) -> bool {
     let ext = path.extension();
-    ext.is_none() || ext == Some(OsStr::new("rs"))
+    ext == Some(OsStr::new("rs")) ||
```
[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/util/toml/mod.rs:80]: > Should we always verify that embedded manifest is a file?

We have a separate, more specific error check for that.

> (Not a blocker) Is it worth having a better error message for each failed case?

What do you have in mind?

[the remark this answers is not in the document]
Comment (weihanglo) [reply src/cargo/util/toml/mod.rs:80]: * When missing file at manifest-path, it shows
    ```
    error: the manifest-path must be a path to a Cargo.toml file
    ```
* When `script.rs` is a directory, it shows
    ```
    error: failed to read `/projects/cargo/test.rs`

    Caused by:
      Is a directory (os error 21)
    ```

[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/util/toml/mod.rs:80]: > When script.rs is a directory, it shows 

This is true for `Cargo.toml` as well.  We can more generally improve it though that is unrelated to this PR

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Thanks for the update!

@bors r+

**Pull request merged**